---
title: "Aggregrate Multiply-genotyped indivs and assign rockfish with rubias"
output: 
  html_notebook:
    toc: true
---




To start off with, let's load data and libs:
```{r load-stuff}
library(tidyverse)
library(CKMRsim)
library(stringr)
library(readxl)

#meta <- readRDS("../data/processed/meta-data-tibble.rds") # not worrying about meta data right now...
genos <- readRDS("../data/processed/called_genos_na_explicit.rds") #%>%
  #filter(NMFS_DNA_ID %in% meta$NMFS_DNA_ID)  # drop those we don't have meta data for
samples <- readRDS("../data/processed/sample-sheet-tibble.rds") #%>%
  #filter(NMFS_DNA_ID %in% meta$NMFS_DNA_ID)
```


## Some initial filters

### Take highest read-depth call for multiply-genotyped DNA_IDs

I'm not sure if there are any of these, but best to leave it in here...
particularly for the re-genotyped fish from Brittany's OSU samples.

Now, here is a harder operation: if an individual is multiply-genotyped, take the
genotype with the highest total read depth.  
```{r take-just-one}
# slow-ish function to get the total read depth column
tdepth <- function(a, d) {
  if(any(is.na(a))) {
    return(NA)
  }
  if(a[1]==a[2]) {
    return(d[1])
  } else {
    return(d[1] + d[2])
  }
  
}
# this takes the highest read-depth instance of each duplicately-genotyped individual.
geno_one_each <- genos %>%
  group_by(NMFS_DNA_ID, locus, gtseq_run, id) %>%
  mutate(total_depth = tdepth(allele, depth)) %>%
  ungroup() %>%
  arrange(NMFS_DNA_ID, locus, total_depth, gtseq_run, id, depth) %>%
  group_by(NMFS_DNA_ID, locus) %>%
  mutate(rank = 1:n()) %>%
  ungroup() %>%
  filter(rank <= 2)
```

### Remove the 6 loci which Hayley has been removing

```{r remove-loci}
# read in a list of the 6 loci
to_remove <- read_csv("../data/loci_to_remove.csv")

# only keep the loci that are not those 6
keepers <- geno_one_each %>%
  anti_join(., to_remove, by = "locus")

# that should leave 90 loci  
```


### Toss out indivs with missing data at more than 15 loci
Now, toss out any individual with fewer than 75 non-missing loci
```{r toss-missers}
no_hi_missers <- keepers %>% 
  group_by(NMFS_DNA_ID) %>%
  filter(sum(!is.na(allele)) >= (75*2))
```
So, we started with `r length(unique(geno_one_each$NMFS_DNA_ID))` 
and after filtering out indivs with fewer than 75 genotyped loci, we were left with 
`r length(unique(no_hi_missers$NMFS_DNA_ID))` individuals.  Those are the ones that
we will run through rubias to identify to species.

## Read in baseline genotypes and remove loci and individuals with too much missing data

```{r read-spp-genos}
# read in genotypes identified to species using rubias
spp <- read_csv("../data/reported_haplotype_SebSppID_11102017.csv")

select_spp <- spp %>%
  select(group, locus, indiv.ID, haplotype.1, haplotype.2)

spp.id <- select_spp %>%
  gather("gene_copy", "allele", 4:5) %>%
  mutate(gene_copy = ifelse(gene_copy == "haplotype.1", 1, 2))

# only keep the loci that are not the 6 removed from the previous dataset
spp.id_loc <- spp.id %>%
  anti_join(., to_remove, by = "locus")
# that should leave 90 loci 


# remove samples with missing data at more than 15 loci (per Hayley's workflow)
spp.id_no_missers <- spp.id_loc %>%
  group_by(indiv.ID) %>%
  filter(sum(!is.na(allele)) >= (75*2))


# add reference column to prepare data for rubias
spp.id_loc1 <- spp.id_no_missers %>%
  mutate(sample_type = "reference")

x <- spp.id_loc1 %>%
  mutate(repunit = group)

# reorder the columns
spp.id1 <- x[,c(6,7,1,3,2,4:5)]

spp.id2 <- spp.id1 %>%
  rename(collection = group) %>%
  rename(indiv = indiv.ID)

# create an artificial ID for the spp.id
#spp.id2 <- spp.id1 %>%
#  unite(NMFS_DNA_ID, 1:2)

# get the data frames into the same format
no_hi_missers2 <- no_hi_missers %>%
  dplyr::select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  rename(indiv = NMFS_DNA_ID) %>%
  mutate(sample_type = "mixture") %>%
  mutate(repunit = NA) %>%
  mutate(collection = "osu_samples")

# reorder

no_hi_missers2[, c(5:7,1:4)]

# combine the data into a single df
alleles <- bind_rows(spp.id2, no_hi_missers2)

alleles
```


We are going to do this by turning alleles into integers and spreading it and then getting it into the right format to run rubias.
```{r spread-genos}
# first make integers of the alleles
alle_idxs <- alleles %>% 
  #dplyr::select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  group_by(locus) %>%
  mutate(alleidx = as.integer(factor(allele, levels = unique(allele)))) %>%
  ungroup() %>%
  arrange(indiv, locus, alleidx) # rubias can handle NA's, so no need to change them to 0's
  
# select just the columns to retain and spread the alleles
alle_idx2 <- alle_idxs[,-7]
  
two_col <- alle_idx2 %>%
  unite(loc, locus, gene_copy, sep = ".") %>%
  spread(loc, alleidx)

two_col

# write this file to a thing that can be read-into other softwares
#two_col %>%
# write_csv("csv_outputs/genos_two_col.csv")
```

Okay, after some reading, it looks like I need to use `infer_mixture` in rubias.
Which require two separate data frames, one with the reference genotypes and the other with the mixture. 

I'll split the data frame that I created, but it needed to be bunged together for the conversion of alleles to integers.

```{r split-frames}
# split up the reference and mixture data frames
sp_mix <- two_col %>%
  filter(sample_type == "mixture")

sp_ref <- two_col %>%
  filter(sample_type == "reference")
```


```{r run-rubias}
# Now that the data are in the corret format, load Rubias
library(rubias)

# perform self-assignment of reference samples
ref_self <- self_assign(sp_ref, gen_start_col = 5)

# and take a quick look at those assignments
ref_self %>%
  filter(inferred_repunit == repunit) %>%
  filter(scaled_likelihood > 0.95)

```

```{r run-mixture}
# perform mixture-assignment on baseline colony samples
mix_assign <- infer_mixture(reference = sp_ref, mixture = sp_mix, gen_start_col = 5, method = "MCMC", reps = 2000, burn_in = 100)

# That was fast, let's take a look
head(mix_assign)

# the individual data is in 
mix_assign$indiv_posteriors %>%
  arrange(desc(log_likelihood))

# are there any rockfish that don't seem like the correct species is in the reference?
osu_assignments <- mix_assign$indiv_posteriors %>%
  filter(PofZ == 1)

osu_assign2 <- osu_assignments[,-10]

#write_csv(osu_assign2, "csv_outputs/osu_assignments.csv")

osu_assign2
```

The posterior means of group membership in each collection is in the PofZ column - there are 1,220 individuals with a PofZ = 1.

```{r combine-brittanys-data-with-rubias-output}
# what I want to do here is join Brittany's sample sheet with the rubias assignments

brit_data <- read_excel("../extdata/brittany_osu_samplesheets/OSURockfish_gtseq_Species_ID_Data.xlsx")

brit_data <- brit_data[1:21]

tmp1 <- osu_assign2 %>%
  left_join(., brit_data, by = c("indiv" = "NMFS_DNA_ID"))

brit_assign <- tmp1[1:21]

# which of brittany's samples still don't have genotype info?
brit_data %>%
  anti_join(., osu_assign2, by = c("NMFS_DNA_ID" = "indiv"))
```

Maybe some of these samples just had a PofZ larger/smaller than 1?
(eventually I will update `brit_data` with all of the appropriate meta data for this project.)

```{r lesser-assignments}
# I think there are individual assignments for each sample to each reference.
# I want to take only the top assignment for each sample.
kept_assignments <- mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(PofZ > 0.95) %>%
  arrange(desc(PofZ))

kept_assignments
```
Using Hayley's threshold of 0.95 for the PofZ, we have 1,435 individuals retained.

Now take a look at which fish from Brittany's samples are included:
```{r brits-assignemnts}
# using the data I read in previously
tmp2 <- kept_assignments %>%
  left_join(., brit_data, by = c("indiv" = "NMFS_DNA_ID"))

brit_assign <- tmp2[1:21]

# which of brittany's samples still don't have genotype info?
brit_data %>%
  anti_join(., kept_assignments, by = c("NMFS_DNA_ID" = "indiv"))

```
That didn't increase the number of samples retained by much... 

Are there really 375 samples that don't have genotypes?

```{r look-at-all-assignments}
all_assign <- mix_assign$indiv_posteriors %>%
  group_by(indiv)

# which of brittany's samples don't have genotype info?
brit_data %>%
  anti_join(., all_assign, by = c("NMFS_DNA_ID" = "indiv"))

```
Okay, those 375 samples just aren't in the data set. They're not just filtered out by the 0.95 PofZ assignment threshold.

```{r compare-brit-assign}
head(brit_data) # here is the "SPECIES" column

kept_assignments %>%
  left_join(., brit_data, by = c("indiv" = "NMFS_DNA_ID")) %>%
  select(1:9,11:30) %>%
  write_csv("csv_outputs/osu_kept_assignments.csv")
  
```


## Repository meta data for Brittany's samples

```{r add-meta-data}
batch_4792 <- read_csv("../extdata/batch4792.csv")
batch_4969 <- read_csv("../extdata/batch4969.csv")

# somehow the structure for the sample IDs is different in the two files...
# change that.
batch_4792$SAMPLE_ID <- as.character(batch_4792$SAMPLE_ID)

# bind those things together
meta <- bind_rows(batch_4969, batch_4792)

```

The `meta` file should contain all the samples for Brittany's project.

```{r meta-assign}
brit_all_assign <- meta %>%
  left_join(., kept_assignments, by = c("NMFS_DNA_ID" = "indiv"))

brit_all_assign %>%
  select(1:5,7:9,17,25:30) %>%
  write_csv("csv_outputs/brittany_estuary_assignments.csv")

# and a list of the samples that were not assigned (for one reason or another)
brit_missing <- brit_all_assign %>%
  anti_join(., kept_assignments, by = c("NMFS_DNA_ID" = "indiv")) %>%
  #group_by(BOX_ID) %>%
  #tally() 
  select(1:30) #%>%
  #write_csv("csv_outputs/brittany_unassigned_samples.csv")
  
```
192 samples. That's still an unfortunate lot.
I wonder if they are the same samples that didn't genotype well last time?

```{r change-threshold}
# if the assignment threshold is lower than 0.95, that might include more samples?
ok_assignments <- mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(PofZ > 0.80) %>%
  arrange(desc(PofZ))

# and join those to Brittany's samples
brit_all_assign %>%
  anti_join(., ok_assignments, by = c("NMFS_DNA_ID" = "indiv"))

```
Still 192 samples!

Of primary concern is whether we actually genotyped these at all. Let's go back to the data.
```{r samples-run}
samples %>%
  anti_join(., brit_all_assign, by = "NMFS_DNA_ID")
# 686 samples that were run that were not brittany's.

brit_all_assign %>%
  anti_join(., samples, by = "NMFS_DNA_ID")
# 0 samples of brittany's that were not run.
# well, at least that's good.

# now check genotypes
brit_all_assign %>%
  anti_join(., genos, by = "NMFS_DNA_ID")
# all of brittany's samples had genotypes that were read in from the rds files.

# where were they kicked out?
hi_missers <- brit_all_assign %>%
  anti_join(., no_hi_missers, by = "NMFS_DNA_ID")

missers_loci <- keepers %>% # keepers is the data frame after removing 6 loci
  inner_join(., hi_missers, by = "NMFS_DNA_ID") %>%
  group_by(NMFS_DNA_ID) %>%
  select(NMFS_DNA_ID, allele) %>%
  tally(is.na(allele)) %>%
  mutate(missing_loci = n/2)

# there shouldn't be any missers with < 15 missing loci
missers_loci %>%
  filter(missing_loci < 15) 
  
# stick that data table onto the 192 samples with too much missing data
missers_loci %>%
  left_join(., brit_missing, by = "NMFS_DNA_ID") %>%
  select(1,3:17,19,21:24) %>%
  write_csv("csv_outputs/OSU_unassigned_missing_data.csv")

```
Yup. The 192 samples are in the hi-missers data - they had lots of missing genotypes.


## Looking at the success of re-genotyped samples

```{r tossers}
# just the samples from gtseq 63
gtseq63 <- samples %>%
  filter(gtseq_run == "gtseq63")

# now add assignments just for those samples
gtseq63 %>%
  left_join(., kept_assignments, by = c("NMFS_DNA_ID" = "indiv")) %>%
  select(1:2, 5:7, 17:22) %>%
  #write_csv("csv_outputs/gtseq63_assignments.csv")
  filter(ssBOX_ID == "R374") #%>%
  #write_csv("csv_outputs/gtseq63_R374_assignments.csv")

```
Here's what Cassie said was in that run:
This run has box R373 and two mixed plates (1 of those mixed plates is partial). 
Plate 1 has the samples from plate R374 that need species ID – Brittany's samples in `1A - 3A`; samples for Helen Killeen in `3B - 4A`; two re-picks from your NFS project in `4B` and `4C`; and ONE tentative Mexican rockfish in `4D`. It also has re-runs for Brittany's project. The two NSF re-picks are highlighted in the attached list of black and yellow sibs that you gave me. We re-extracted these since there was tissue remaining and we also re-ran the original DNA extracts in this run. 
Plate 2rr are all re-runs including your NSF black and yellow sibs and the rest of the failed samples from Bailey's run for Brittany's project. 

```{r easier-format}
gtseq63 %>%
  filter(ssBOX_ID == "R282")
```

```{r re-runs}
# which samples from gtseq63 don't have kept assignments?
gtseq63 %>%
  anti_join(., kept_assignments, by = c("NMFS_DNA_ID" = "indiv")) %>%
  filter(ssBOX_ID == "R133")

```
All the samples from R373 were not re-runs. (23 samples)
The 4 samples that failed from R374 were for Helen Killeen.

47 -27 = 20 samples that failed from the re-runs? Was that out of 108 total?

And if we did re-run all of Brittany's fish from previous runs, is the majority of the missing data from the new samples? 

