---
title: "assign non-target gopher-black-and-yellow rockfish with rubias"
output: html_notebook
---

To start off with, let's load data and libs:
```{r load-stuff}
library(tidyverse)
library(CKMRsim)
library(stringr)
library(readxl)
library(rubias)

meta <- readRDS("../../nsf-kelp-rockfish-ckmr/data/processed/meta-data-tibble.rds") %>%
  filter(REPORTED_LIFE_STAGE != "Extrusion Larvae") %>% # remove the extrustion larvae
  filter(REPORTED_LIFE_STAGE != "Adult") %>% # these are mothers for the larvae
  filter(REPORTED_LIFE_STAGE != "ADULT") %>% # just the juveniles for this analysis
  filter(SPECIES != "melanops") # no black rockfish for this.

genos <- readRDS("../nsf_data/processed/called_genos_na_explicit.rds") %>%
  filter(NMFS_DNA_ID != "Morin") # remove the data for Phil Morin
samples <- readRDS("../nsf_data/processed/sample-sheet-tibble.rds") %>%
  filter(NMFS_DNA_ID %in% meta$NMFS_DNA_ID)
```

## Some initial filters

### Take highest read-depth call for multiply-genotyped DNA_IDs

I'm not sure if there are any of these, but best to leave it in here...
if an individual is multiply-genotyped, take the
genotype with the highest total read depth.  
```{r take-just-one}
# slow-ish function to get the total read depth column
tdepth <- function(a, d) {
  if(any(is.na(a))) {
    return(NA)
  }
  if(a[1]==a[2]) {
    return(d[1])
  } else {
    return(d[1] + d[2])
  }
  
}
# this takes the highest read-depth instance of each duplicately-genotyped individual.
geno_one_each <- genos %>%
  group_by(NMFS_DNA_ID, locus, gtseq_run, id) %>%
  mutate(total_depth = tdepth(allele, depth)) %>%
  ungroup() %>%
  arrange(NMFS_DNA_ID, locus, total_depth, gtseq_run, id, depth) %>%
  group_by(NMFS_DNA_ID, locus) %>%
  mutate(rank = 1:n()) %>%
  ungroup() %>%
  filter(rank <= 2)
```


```{r}
# how many sample genotypes now?
geno_one_each %>%
  inner_join(., samples, by = "NMFS_DNA_ID") %>%
  select(NMFS_DNA_ID) %>%
  unique()
```
```{r}
# which means duplicate samples for how many?
samples %>%
  group_by(NMFS_DNA_ID) %>%
  tally() %>%
  arrange(desc(n))
```
12 fish genotyped 2x.

### Remove the 6 loci which Hayley has been removing

```{r remove-loci}
# read in a list of the 6 loci
to_remove <- read_csv("../data/loci_to_remove.csv")

# only keep the loci that are not those 6
keepers <- geno_one_each %>%
  anti_join(., to_remove, by = "locus")
```
that should leave 90 loci  

```{r how-many-now}
unique_ids <- samples %>%
  select(NMFS_DNA_ID) %>%
  unique()

keepers <- keepers %>%
  inner_join(., unique_ids)

```


### Toss out indivs with data at fewer than 25 loci
Now, toss out any individual with fewer than 25 non-missing loci
```{r toss-missers}
no_hi_missers <- keepers %>% 
  group_by(NMFS_DNA_ID) %>%
  filter(sum(!is.na(allele)) >= (25*2))
```
So, we started with `r length(unique(geno_one_each$NMFS_DNA_ID))` 
and after filtering out indivs with fewer than 75 genotyped loci, we were left with 
`r length(unique(no_hi_missers$NMFS_DNA_ID))` individuals.  Those are the ones that
we will run through rubias to identify to species.

## Read in baseline genotypes and remove loci and individuals with too much missing data

```{r read-spp-genos}
# read in genotypes identified to species using rubias
spp <- read_csv("../data/reported_haplotype_SebSppID_11102017.csv")

select_spp <- spp %>%
  select(group, locus, indiv.ID, haplotype.1, haplotype.2)

spp.id <- select_spp %>%
  gather("gene_copy", "allele", 4:5) %>%
  mutate(gene_copy = ifelse(gene_copy == "haplotype.1", 1, 2))

# only keep the loci that are not the 6 removed from the previous dataset
spp.id_loc <- spp.id %>%
  anti_join(., to_remove, by = "locus")
# that should leave 90 loci 
```

From the .csv output from microhaplot, we need to reformat the data for rubias.
```{r format-reference-for-rubias}
# add reference column to prepare data for rubias
spp.id_loc1 <- spp.id_loc %>%
  mutate(sample_type = "reference")

# add the repunit column that is identical to the group in this case
x <- spp.id_loc1 %>%
  mutate(repunit = group)

# reorder the columns and get the data frame in the right format
spp.id1 <- x[,c(6,7,1,3,2,4:5)]
spp.id2 <- spp.id1 %>%
  rename(collection = group) %>%
  rename(indiv = indiv.ID)
```

To ensure that haplotypes are coded identically across both the reference and mixture genotypes, we need to combine the two data sets, and then change haplotypes into numeric alleles.
```{r format-the-mixture}
# get the reference and mixture data frames into the same format
no_hi_missers2 <- no_hi_missers %>%
  dplyr::select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  rename(indiv = NMFS_DNA_ID) %>%
  mutate(sample_type = "mixture") %>%
  mutate(repunit = NA) %>%
  mutate(collection = "nsf_samples")

# reorder the columns for the mixture data frame
no_hi_missers3 <- no_hi_missers2[, c(5:7,1:4)]

# combine both data sets into a single df
alleles <- bind_rows(spp.id2, no_hi_missers3)

```


Here we turn alleles into integers, spread the data frame, and then get it into the right format to run rubias:
```{r spread-genos}
# first make integers of the alleles
alle_idxs <- alleles %>% 
  #select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  group_by(locus) %>%
  mutate(alleidx = as.integer(factor(allele, levels = unique(allele)))) %>%
  ungroup() %>%
  arrange(indiv, locus, alleidx) # rubias can handle NA's, so no need to change them to 0's
  
# select just the columns to retain and spread the alleles
alle_idx2 <- alle_idxs[,-7]
  
two_col <- alle_idx2 %>%
  unite(loc, locus, gene_copy, sep = ".") %>%
  spread(loc, alleidx)

# write this file to a thing that can be read-into other softwares
#two_col %>%
# write_csv("csv_outputs/genos_two_col.csv")
```

We will use `infer_mixture` in rubias, which requires two separate data frames, one with the reference genotypes and the other with the mixture. 

I'll split the data frame that I created (but it needed to be bunged together for the conversion of alleles to integers).
```{r split-frames}
# split up the reference and mixture data frames
sp_mix <- two_col %>%
  filter(sample_type == "mixture")

# sp_mix %>%
#   write_csv("csv_outputs/nsf-mixture-two-col-rubias.csv") # this was for the PCA

sp_ref <- two_col %>%
  filter(sample_type == "reference")

# how many gopher v. black-and-yellow rockfishes are in the reference file?
sp_ref %>%
  filter(repunit %in% c("Scarnatus", "Schrysomelas")) %>% # 350 combined.
  filter(repunit == "Schrysomelas") # 110 of them are black-and-yellow.
```

Experiment with equalizing the number of gopher/black-and-yellow rockfish in the reference file. Let's do 110 of each.
```{r equalize-ref-gby}
gopher_ref <- sp_ref %>%
  filter(repunit == "Scarnatus") %>%
  sample_n(110, replace = FALSE)

# now create a reference that removes gopher from the reference and then put back just the 110 we want
spp_reference <- sp_ref %>%
  filter(repunit != "Scarnatus") %>%
  bind_rows(., gopher_ref)

```


Now, with the two data frames separated, run the `infer mixture` analysis:
```{r run-mixture}
# perform mixture-assignment on baseline colony samples
mix_assign <- infer_mixture(reference = spp_reference, mixture = sp_mix, gen_start_col = 5, method = "MCMC", reps = 2000, burn_in = 100)

# That was fast, let's take a look
head(mix_assign)

# the individual data is in 
mix_assign$indiv_posteriors %>%
  arrange(desc(log_likelihood))

# look at the fish that were assigned with > 50% confidence
mix_assign$indiv_posteriors %>%
  filter(PofZ > 0.5) %>%
  group_by(repunit) %>%
  tally() %>%
  arrange(desc(n)) %>%
  write_csv("csv_outputs/nsf-PofZ-50.csv")
 
# What about at higher confidence?
# PofZ > 0.99
mix_assign$indiv_posteriors %>%
  filter(PofZ > 0.99) %>%
  group_by(repunit) %>%
  tally() %>%
  arrange(desc(n)) %>%
  write_csv("csv_outputs/nsf-PofZ-99.csv")

```
The posterior means of group membership in each collection is in the PofZ column.

Make a quick little plot to display the diversity of species in our NSF juvenile samples.
```{r quick-plot}
# ordered by number of samples
# ggplot(grp_tally2, (aes(x = reorder(repunit, -n), y = n))) + 
#   geom_bar(stat = "identity") +
#   theme_bw() +
#   ylab("Samples") +
#   xlab("Species") +
#   theme(
#     axis.text.x = element_text(angle = 45, size = 16),
#     axis.text.y = element_text(size = 16),
#     axis.title = element_text(size = 18)) +
#   scale_color_brewer(palette = "Set1")
# 
#   
# ggsave("pdf_outputs/nsf_juv_id.pdf", height = 7, width = 12)
```

```{r}
# Now just the non-KGBC
# grp_tally2 %>%
#   filter(repunit != "Satrovirens" & repunit != "Scarnatus" & repunit != "Schrysomelas" & repunit !="Scaurinus") %>%
#   ggplot((aes(x = reorder(repunit, -n), y = n))) + 
#   geom_bar(stat = "identity") +
#   theme_bw() +
#   ylab("Samples") +
#   xlab("Species") +
#   theme(
#     axis.text.x = element_text(angle = 45, size = 16),
#     axis.text.y = element_text(size = 16),
#     axis.title = element_text(size = 18)) +
#   scale_color_brewer(palette = "Set1")
# 
# 
# ggsave("pdf_outputs/non_kgbc_juv_id.pdf", height = 7, width = 10)
# 
# 
# grp_tally2 %>%
#   filter(repunit != "Satrovirens" & repunit != "Scarnatus" & repunit != "Schrysomelas" & repunit !="Scaurinus") %>% ungroup %>%
#   arrange(desc(n))



# maybe I could color the bars if they are part of the pteropodus family?
# ptero <- c("Satrovirens", "Scarnatus", "Scaurinus", "Schrysomelas", "Snebulosus")
# 
# ptero
```


```{r lesser-assignments}
# There are individual assignments for each sample to each reference.
# If I want to keep the top assignment for each sample:
kept_assignments <- mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(PofZ > 0.5) %>%
  arrange(desc(PofZ))

kept_assignments %>%
  select(-missing_loci) %>%
  write_csv("csv_outputs/nsf-PofZ-0.5_assignments.csv")
```

## Picking out the gopher, black-and-yellow, and copper rockfishes

Interesting to see all the other species in there - mostly china and quillback. (And in reality, these non-target species only make up 1-2% of samples.)

But what we actually want is a list of NMFS IDs for the fish assigned to gopher, black-and-yellow, and copper, which we can then use to select those fish from our NSF dataset, where we presumably have many more alleles for those species because the VCF used to call genotypes included hundreds of samples for each of those species.

This dataset includes 12,002 samples.
```{r}
 mix_assign$indiv_posteriors %>%
  #group_by(indiv) %>%
  filter(PofZ > 0.99) %>% 
  filter(collection %in% c("Scarnatus", "Schrysomelas")) %>%
  select(-missing_loci) %>%
  rename(NMFS_DNA_ID = indiv) %>%
  write_csv("csv_outputs/nsf-GBY-PofZ99.csv")
  
```

And for coppers, a separate list.
```{r}
mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(PofZ > 0.5) %>% # keep all potential assignments to check again with GSI SIM in the nsf workflow
  filter(collection == "Scaurinus") %>%
  select(indiv) %>%
  rename(NMFS_DNA_ID = indiv) %>%
  write_csv("csv_outputs/nsf-copper-samples-ids.csv")

```

### Check the assignment probabilities of the new pairs

When I reanalyzed the gopher/black-and-yellow pairs without the gsi sim > 99 score threshold, I found an additional 5 matches. Since we know they are not non-target species thanks to the rubias species identification analysis, I want to look at what their PofZ scores before including them in my results.

```{r}
# read in the new matches
matches <- read_csv("../../nsf-gopher-black-and-yellow/Rmd/csv_outputs/extra_gby_sibs.csv")
```


```{r}
sp_assignments <- mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(PofZ > 0.5) 
```


```{r}
matches %>%
  left_join(., sp_assignments, by = c("name1" = "indiv")) %>%
  left_join(., sp_assignments, by = c("name2" = "indiv")) %>%
  select(name1, name2, repunit.x, PofZ.x, repunit.y, PofZ.y)

```
Of these five matches, there are three that come up as gopher-gopher pretty solidly. Furthermore, two of those three they have a PofZ for each individual > 0.99.

Those are in.

So for the remaining three matches, I think I might be inclined to include the one that is gopher-gopher with one individual with a PofZ > 0.90, but not the two matches that include one gopher and one black-and-yellow sib?

Let me look at the overall PofZ for those samples to make sure it's just gopher/black-and-yellow
```{r}
mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(indiv == "R026051") %>%
  arrange(desc(PofZ)) %>%
  filter(PofZ > 0.01)
```
For that sample, the assignment is to gopher/black-and-yellow. 
What about the next one?
```{r}
mix_assign$indiv_posteriors %>%
  group_by(indiv) %>%
  filter(indiv == "R026909") %>%
  arrange(desc(PofZ)) %>%
  filter(PofZ > 0.01)

```
Same with this one.


